tsl/src/bgw_policy/job.c

```diff
@@ -592,24 +647,60 @@ policy_recompression_read_and_validate_config(Jsonb *config, PolicyCompressionDa
 bool
 policy_recompression_execute(int32 job_id, Jsonb *config)
 {
-	int32 chunkid;
+	List *chunkid_lst;
+	ListCell *lc;
 	const Dimension *dim;
 	PolicyCompressionData policy_data;
+	bool distributed, used_portalcxt = false;
+	MemoryContext saved_cxt, multitxn_cxt;
 
 	policy_recompression_read_and_validate_config(config, &policy_data);
 	dim = hyperspace_get_open_dimension(policy_data.hypertable->space, 0);
-	chunkid = get_chunk_to_recompress(dim, config);
+	distributed = hypertable_is_distributed(policy_data.hypertable);
+	/* we want the chunk id list to survive across transactions. So alloc in
+	 * a different context
+	 */
+	if (PortalContext)
+	{
+		/*if we have a portal context use that - it will get freed automatically*/
+		multitxn_cxt = PortalContext;
+		used_portalcxt = true;
+	}
+	else
+	{
+		/* background worker job does not go via usual CALL path, so we do
+		 * not have a PortalContext */
+		multitxn_cxt =
+			AllocSetContextCreate(TopMemoryContext, "CompressionJobCxt", ALLOCSET_DEFAULT_SIZES);
+	}
+	saved_cxt = MemoryContextSwitchTo(multitxn_cxt);
+	chunkid_lst = get_chunk_to_recompress(dim, config);
+	MemoryContextSwitchTo(saved_cxt);
 
-	if (chunkid == INVALID_CHUNK_ID)
+	if (!chunkid_lst)
+	{
 		elog(NOTICE,
 			 "no chunks for hypertable \"%s.%s\" that satisfy recompress chunk policy",
 			 policy_data.hypertable->fd.schema_name.data,
 			 policy_data.hypertable->fd.table_name.data);
-
-	if (chunkid != INVALID_CHUNK_ID)
+		ts_cache_release(policy_data.hcache);
+		if (!used_portalcxt)
+			MemoryContextDelete(multitxn_cxt);
+		return true;
+	}
+	ts_cache_release(policy_data.hcache);
+	if (ActiveSnapshotSet())
+		PopActiveSnapshot();
```

Comment:
There can be several active snapshots. Why do you pop only the top one? Is this intentional?

Reply:
We come here either via CALL or as part of the bgw job - in either case, there is only 1 active snapshot

Reply:
We could add an assertion afterwards:
```
Assert(!ActiveSnapshotSet());
```